File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/buildfilter/BuildFilterResult.java
package net.nemerosa.ontrack.model.buildfilter;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;

@Data
@AllArgsConstructor(access = AccessLevel.PROTECTED)
public class BuildFilterResult {

    private final boolean accept;
    private final boolean goingOn;

    public static BuildFilterResult stopNowIf(boolean condition) {
        return new BuildFilterResult(!condition, !condition);
    }

    public static BuildFilterResult stopNow() {
        return new BuildFilterResult(false, false);
    }

    public static BuildFilterResult ok() {
        return new BuildFilterResult(true, true);
    }

    public static BuildFilterResult notAccept() {
        return new BuildFilterResult(false, true);
    }

    public static BuildFilterResult accept() {
        return new BuildFilterResult(true, true);
    }

    public BuildFilterResult goingOn() {
        return new BuildFilterResult(accept, true);
    }

    public BuildFilterResult stop() {
        return new BuildFilterResult(accept, false);
    }

    public BuildFilterResult acceptIf(boolean condition) {
        return new BuildFilterResult(accept && condition, goingOn);
    }

    public BuildFilterResult goOnIf(boolean condition) {
        return new BuildFilterResult(accept, goingOn && condition);
    }

    public BuildFilterResult doAccept() {
        return new BuildFilterResult(true, goingOn);
    }
}

File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/buildfilter/BuildFilterResult.kt
package net.nemerosa.ontrack.model.buildfilter


data class BuildFilterResult private constructor(
    val accept: Boolean = false,
    val goingOn: Boolean = false,
) {

    fun goingOn(): BuildFilterResult {
        return BuildFilterResult(accept, true)
    }

    fun stop(): BuildFilterResult {
        return BuildFilterResult(accept, false)
    }

    fun acceptIf(condition: Boolean): BuildFilterResult {
        return BuildFilterResult(accept && condition, goingOn)
    }

    fun goOnIf(condition: Boolean): BuildFilterResult {
        return BuildFilterResult(accept, goingOn && condition)
    }

    fun doAccept(): BuildFilterResult {
        return BuildFilterResult(true, goingOn)
    }

    companion object {
        fun stopNowIf(condition: Boolean): BuildFilterResult {
            return BuildFilterResult(!condition, !condition)
        }

        fun stopNow(): BuildFilterResult {
            return BuildFilterResult(false, false)
        }

        fun ok(): BuildFilterResult {
            return BuildFilterResult(true, true)
        }

        fun notAccept(): BuildFilterResult {
            return BuildFilterResult(false, true)
        }

        fun accept(): BuildFilterResult {
            return BuildFilterResult(true, true)
        }
    }
}
